### PR TITLE
server,session: Add status vars for compression

### LINKED ISSUE #44473: Support `SHOW SESSION STATUS LIKE "Compression";`
## Feature Request

Since https://github.com/pingcap/tidb/pull/36780 has been merged to master branch, `tiup playground nightly` supports MySQL compressed protocol.

To see if the compression is actually enabled or not, `SHOW SESSION STATUS LIKE "Compression";` can be used for MySQL which is not available for TiDB yet. Refer to https://stackoverflow.com/questions/27427041/checking-connection-compression-usage

**Describe the feature you'd like:**
I wish TiDB support `SHOW SESSION STATUS LIKE "Compression";` as MySQL does. Here are TiDB and MySQL behaviors.

--- TiDB
```sql
$ tiup playground nightly
$ mysql --comments --host 127.0.0.1 --port 4000 -u root --compress
mysql> SHOW SESSION STATUS LIKE "Compression";
Empty set (0.02 sec)

mysql>
mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v7.2.0-alpha
Edition: Community
Git Commit Hash: be8d86504402cf6e9648fa4972d29e6ac3f9e9a8
Git Branch: heads/refs/tags/v7.2.0-alpha
UTC Build Time: 2023-06-05 14:25:17
GoVersion: go1.20.3
Race Enabled: false
Check Table Before Drop: false
Store: tikv
1 row in set (0.00 sec)

mysql>
```

--- MySQL
```sql
% mysql --comments --host 127.0.0.1 --port 3306 -u root --compress

WARNING: --compress is deprecated and will be removed in a future version. Use --compression-algorithms instead.
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 8
Server version: 8.0.33 Homebrew

Copyright (c) 2000, 2023, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> SHOW SESSION STATUS LIKE "compression";
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| Compression   | ON    |
+---------------+-------+
1 row in set (0.01 sec)

mysql> 
```

**Describe alternatives you've considered:**

`\status` is available for both TiDB and MySQL by cheking `Protocol:		Compressed`. However, this feature is only available if the connection is made by mysql cli, not other application access like PHP with `PDO::MYSQL_ATTR_COMPRESS`.

```sql
% mysql --comments --host 127.0.0.1 --port 4000 -u root --compress

WARNING: --compress is deprecated and will be removed in a future version. Use --compression-algorithms instead.
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 407
Server version: 8.0.11-TiDB-v7.2.0-alpha TiDB Server (Apache License 2.0) Community Edition, MySQL 8.0 compatible

Copyright (c) 2000, 2023, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> \status
--------------
mysql  Ver 8.0.33 for macos13.3 on x86_64 (Homebrew)

Connection id:		407
Current database:
Current user:		root@127.0.0.1
SSL:			Cipher in use is TLS_AES_128_GCM_SHA256
Current pager:		stdout
Using outfile:		''
Using delimiter:	;
Server version:		8.0.11-TiDB-v7.2.0-alpha TiDB Server (Apache License 2.0) Community Edition, MySQL 8.0 compatible
Protocol version:	10
Connection:		127.0.0.1 via TCP/IP
Server characterset:	utf8mb4
Db     characterset:	utf8mb4
Client characterset:	utf8mb4
Conn.  characterset:	utf8mb4
TCP port:		4000
Protocol:		Compressed
Binary data as:		Hexadecimal
Uptime:			3 min 2 sec

Threads: 0  Questions: 0  Slow queries: 0  Opens: 0  Flush tables: 0  Open tables: 0  Queries per second avg: 0.000
--------------
```

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

Issue Number: close #44473 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
The `Compression`, `Compression_algorithm` and `Compression_level` status variables were added.
```

### CODE REVIEW

**Comment by dveeden** [comment]
This is how this now looks:
```
$ mysqlsh --compression-algorithms=zlib --sql mysql://root@127.0.0.1:4000/test

sql> SHOW SESSION STATUS LIKE 'Compression%';
+-----------------------+-------+
| Variable_name         | Value |
+-----------------------+-------+
| Compression           | ON    |
| Compression_algorithm | zlib  |
| Compression_level     | 6     |
+-----------------------+-------+
3 rows in set (0.0080 sec)
```

**Comment by dveeden** [comment]
The MySQL Client show this, both with and without this PR:
```
mysql> \s
...
Protocol:		Compressed, algorithms: zlib, zstd level: 3
...
```

And MySQL Shell with this PR:
```
sql> \s
...
Compression:                  Enabled (zstd)
...
```

And MySQL Shell without this PR:
```
sql> \s
...
Compression:                  Enabled
...
```

**Review by djshow832 — APPROVED** [review]

File: pkg/server/conn.go [hunk 1]
```diff
@@ -2581,3 +2599,41 @@ func (cc *clientConn) ReadPacket() ([]byte, error) {
...
+	case mysql.CompressionZlib:
+		m[statusCompression] = "ON"
+		m[statusCompressionAlgorithm] = "zlib"
+		m[statusCompressionLevel] = mysql.ZlibCompressDefaultLevel
+	case mysql.CompressionZstd:
+		m[statusCompression] = "ON"
+		m[statusCompressionAlgorithm] = "zstd"
+		m[statusCompressionLevel] = vars.CompressionLevel
+	default:
+		logutil.Logger(context.Background()).Debug(
```
Comment (lance6716) [inline pkg/server/conn.go:2624]: seems it's equal to log.L()

File: pkg/server/conn.go [hunk 2]
```diff
@@ -2581,3 +2599,41 @@ func (cc *clientConn) ReadPacket() ([]byte, error) {
 func (cc *clientConn) Flush(ctx context.Context) error {
 	return cc.flush(ctx)
 }
+
+// Stats returns the connection statistics.
+func (*clientConn) Stats(vars *variable.SessionVars) (map[string]interface{}, error) {
+	m := make(map[string]interface{}, len(clientConnStatus))
+	for name, v := range clientConnStatus {
```
Comment (lance6716) [inline pkg/server/conn.go:2606]: I'm not very sure about the usage of `clientConnStatus`. After `variable.RegisterStatistics(cc)` the scope part of  `clientConnStatus` is not used, can we remove `clientConnStatus`?

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: pkg/server/conn.go (hunk 1 above)
Comment (dveeden) [reply pkg/server/conn.go:2624]: Like this?

File: pkg/server/conn.go (hunk 2 above)
Comment (dveeden) [reply pkg/server/conn.go:2606]: Not sure exactly what you mean with this. Could you explain it a bit more?

File: pkg/server/conn.go (hunk 1 above)
Comment (dveeden) [reply pkg/server/conn.go:2624]: Or something like this?
```patch
diff --git a/pkg/server/conn.go b/pkg/server/conn.go
index 4518be086b..891fba09b3 100644
--- a/pkg/server/conn.go
+++ b/pkg/server/conn.go
@@ -2621,7 +2621,7 @@ func (*clientConn) Stats(vars *variable.SessionVars) (map[string]interface{}, er
                m[statusCompressionAlgorithm] = "zstd"
                m[statusCompressionLevel] = vars.CompressionLevel
        default:
-               logutil.Logger(context.Background()).Debug(
+               logutil.BgLogger().Debug(
                        "unexpected compression algorithm value",
                        zap.Int("algorithm", vars.CompressionAlgorithm),
                )
```

Comment (lance6716) [reply pkg/server/conn.go:2624]: Oh I think the two changes are both OK

Comment (dveeden) [reply pkg/server/conn.go]: Pushed in 5aca0b2f7ce6d89c2e5dcb0e4b581a02e75dbbfa

File: pkg/server/conn.go (hunk 2 above)
Comment (lance6716) [reply pkg/server/conn.go:2606]: I think we can remove `clientConnStatus`.

```diff
diff --git a/pkg/server/conn.go b/pkg/server/conn.go
index 4518be086b..58a9e24942 100644
--- a/pkg/server/conn.go
+++ b/pkg/server/conn.go
@@ -120,12 +120,6 @@ var (
        statusCompressionLevel     = "Compression_level"
 )

-var clientConnStatus = map[string]*variable.StatusVal{
-       statusCompression:          {Scope: variable.ScopeSession, Value: "OFF"},
-       statusCompressionAlgorithm: {Scope: variable.ScopeSession, Value: ""},
-       statusCompressionLevel:     {Scope: variable.ScopeSession, Value: 0},
-}
-
 // newClientConn creates a *clientConn object.
 func newClientConn(s *Server) *clientConn {
        cc := &clientConn{
@@ -2602,10 +2596,7 @@ func (cc *clientConn) Flush(ctx context.Context) error {

 // Stats returns the connection statistics.
 func (*clientConn) Stats(vars *variable.SessionVars) (map[string]interface{}, error) {
-       m := make(map[string]interface{}, len(clientConnStatus))
-       for name, v := range clientConnStatus {
-               m[name] = v.Value
-       }
+       m := make(map[string]interface{}, 3)

        switch vars.CompressionAlgorithm {
        case mysql.CompressionNone:
```

Comment (dveeden) [reply pkg/server/conn.go]: Done in 6081df6cf805cb3df553e26471d47ccc714bbb0f

**Review by lance6716 — APPROVED** [review]

**Review by yudongusa — APPROVED** [review]
LGTM

**Pull request merged**